src/main/java/bots/Telegram.java

```diff
@@ -0,0 +1,336 @@
+
+        novel.downloadChapters();
+        novel.output();
+        if (!novel.failedChapters.isEmpty()) {
+            novelly.execute(new SendMessage(chatId, "Retrying failed chapters..."));
+            novel.retry();
+            if (!novel.failedChapters.isEmpty()) {
+                novelly.execute(new SendMessage(chatId, "Remaining failed chapters: " + novel.failedChapters.size()));
+            }
+        }
+
+        // Send file
+        File epub = new File(novel.saveLocation+"/"+novel.epubFilename);
+        if(epub.exists()) {
+            novelly.execute(new SendDocument(chatId, epub));
+            GrabberUtils.info("EPUB sent: " + novel.epubFilename);
+            // Update downloaded chapter amount
+            userChapterCount.put(userId, chapterDownloadedToday + novel.successfulChapters.size());
+        } else {
+            novelly.execute(new SendMessage(chatId, "Sorry. Could not download the novel."));
+            GrabberUtils.err("EPUB not downloaded: " + messageTxt);
+        }
+    }
+
+    public void updateProgress(long chatId, int currChapter, int lastChapter) {
+        int messageId = (int) downloadMsgIds.get(chatId);
+        novelly.execute(new EditMessageText(chatId, messageId, "Progress: "+(currChapter+1)+"/"+lastChapter));
+    }
+
+    public static void log(String msg) {
+        String time = ZonedDateTime.now().truncatedTo(ChronoUnit.SECONDS).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
+        try {
+            Files.createDirectories(Paths.get(telegramDir));
+            try (BufferedWriter writer = new BufferedWriter(new FileWriter(telegramDir + "/log.txt", true))) {
+                writer.write("[" + time + "] " + msg);
+                writer.write("\n");
+            }
+        } catch (IOException e) {
+            GrabberUtils.err(e.getMessage(), e);
+        }
+    }
+
+    private static String getStringFromFile(String fileName) {
+        StringBuilder resultStringBuilder = new StringBuilder();
+        try (BufferedReader br = new BufferedReader(new FileReader(telegramDir+"/"+fileName))) {
+            String line;
+            while ((line = br.readLine()) != null) {
+                resultStringBuilder.append(line).append("\n");
+            }
+        } catch (IOException e) {
+            GrabberUtils.err("File found: "+ telegramDir+"/"+fileName);
+        }
+        return resultStringBuilder.toString();
+    }
+
+    public void sendMsg(long chatId, String msg) {
+        novelly.execute(new SendMessage(chatId, msg));
+    }
+
+    private String getSourcesString() {
+        StringBuilder sources = new StringBuilder();
+        for(Source source: GrabberUtils.getSources()) {
+            sources.append("[" + source.getName() + "](" + source.getUrl() + ")\n");
+        }
+        return sources.toString();
+    }
+
+    private String getLimitString(int userId) {
+        String msgMaxChNovel = String.format("Max. chapters per novel: %s \n",
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.